### ISSUE TITLE
Add a device= option for GPU/MPS training in npe()/nle()

### ISSUE THREAD

**Issue by pedroliman** [pedroliman/neuralsbi#82]
## Goal

Let users train the neural density estimator on a GPU (CUDA on Linux/Windows, MPS on Mac), with a single portable option that also stays on CPU by default. Motivating use case: train an SBI model as fast as possible on an Apple-silicon Mac.

## Can users do this today without a package change?

No, not reliably. `torch` exposes `with_device()` / `local_device()`, which set the default device for *some* tensor constructors, but our training path mixes devices and crashes.

Wrapping `npe(...)` in `with_device(device = torch_device("mps"), { ... })` fails with:

```
Error: Expected all tensors to be on the same device, but found at least two devices, mps:0 and cpu!
```

Root cause: `with_device()` moves `nn_*` module parameters and `torch_zeros()`/`torch_randn()` to the default device, but `torch_tensor()` built from an R matrix ignores the default device and stays on CPU. Confirmed:

```r
with_device(device = torch_device("mps"), {
  torch_tensor(matrix(1:6,2,3), dtype=torch_float())$device  # cpu   <- ignores default
  torch_zeros(c(2,3))$device                                 # mps:0
  nn_linear(3,4)$parameters[[1]]$device                      # mps:0
})
```

Our training data tensors are built with `torch_tensor(theta/x)` (in `R/train.R`, and in the `de_log_prob`/`de_sample` methods), so they land on CPU while the net lands on the GPU. There are also unconditional `torch_zeros(...)` / `torch_tensor(rev(...))` / `torch_randn(...)` inside `maf_forward`, `maf_inverse`, `mdn_build_tril`, and the NSF equivalents that do *not* take the device from their input tensors. So even fixing the data tensors is not enough; the estimators create scratch tensors on CPU mid-forward.

## Proposed change

Add a `device` argument to `npe()` / `nle()` (default `"cpu"`), thread it into `train_conditional_de()`, and:

1. Create training/validation tensors on `device` in `R/train.R`.
2. `net$to(device = device)` after `build_net()`.
3. Make the estimator forward/inverse code device-safe: derive scratch tensors from an existing tensor's device (e.g. `torch_zeros(..., device = theta\$device)`, `theta\$new_zeros(...)`), rather than the implicit CPU default. Sites: `maf_forward`/`maf_inverse` (`R/flows.R`), `mdn_build_tril` (`R/mdn.R`), the NSF spline code (`R/nsf.R`), and the `de_log_prob`/`de_sample` S3 methods (which must move `x`/`theta` to the net's device and results back to CPU).
4. Store the device on the fit and reuse it (or fall back to CPU) at posterior/sample time. Note serialization: a torch fit already doesn't survive `saveRDS()`; reload via `load_npe()` should land on CPU or a user-chosen device.
5. Validate the device string portably: `"cuda"` only if `cuda_is_available()`, `"mps"` only if `backends_mps_is_available()`, else warn and fall back to CPU. This keeps one code path working on Mac, Windows, and Linux.

## Profiling: is MPS actually faster here?

Benchmarked on this Mac (Apple silicon, `torch` 0.16.0, `cuda_is_available() = FALSE`, `mps` available).

MAF training as configured by the SIR example (theta dim 2, x dim 15, hidden 50x50, 5 transforms, batch 200, n=8000): **CPU ~0.5 s/epoch**. MPS could not be measured end to end because of the mixed-device crash above.

To gauge MPS's ceiling independent of the port, a plain MLP train step (the favorable case for a GPU):

| input dim | hidden | n | batch | CPU | MPS | speedup |
|---|---|---|---|---|---|---|
| 15 | 50 | 8000 | 200 | 0.65s | 2.14s | 0.30x (MPS 3x slower) |
| 15 | 512 | 8000 | 200 | 1.28s | 1.72s | 0.74x (MPS slower) |
| 100 | 1024 | 20000 | 512 | 3.52s | 2.40s | 1.47x (MPS faster) |

Takeaway: at the small net/batch sizes typical of SBI examples like the SIR model, **MPS is slower than CPU** because per-kernel launch overhead dominates; the MAF's Python-loop-over-transforms/dimensions makes this worse (many tiny ops). MPS only wins once the net and batch are large (wide hidden layers, big batches, high-dim x). CUDA on a discrete GPU would likely have a lower crossover point.

## Recommendation

Two parts, worth splitting into PRs:

- **Portability/correctness:** add the `device=` plumbing so GPU training is *possible* and correct on all three platforms. Default stays `"cpu"`.
- **Speed for the Mac use case:** for the small-model regime, device offload is the wrong lever. Bigger wins likely come from cutting per-op overhead in the training loop (larger `batch_size`, vectorizing the MADE transform loop, fewer host syncs like `loss\$item()` every batch) and from the simulation stage (`future` plan). Document that `device = "mps"/"cuda"` helps only for large estimators/batches, so users don't reach for it and get a slowdown.

**Comment by pedroliman** [comment]
## How Python `sbi` handles device selection (for reference)

Checked the upstream `sbi` source. It deliberately keeps CPU as the default and makes GPU explicit, rather than auto-selecting a GPU when one is present.

- **Default is `device = "cpu"`.** The NPE/PosteriorEstimator trainer signature is literally `device: str = "cpu"`. No auto-GPU.
- GPU is **opt-in** via a device string, resolved by `process_device()` (`sbi/utils/torchutils.py`):
  - `None` / `"cpu"` -> CPU. CPU skips the allocation check on purpose, so it does not perturb the RNG state in seeded runs.
  - `"gpu"` -> auto-select in order CUDA -> MPS, and **raises** `RuntimeError` if neither is available. No silent fallback to CPU.
  - `"cuda"` / `"mps"` -> canonicalized (`"cuda"` -> `"cuda:0"`) and validated by actually allocating `torch.randn(1, device=device)`, raising on failure.
  - If the resolved device is MPS, a **one-time warning** fires that PyTorch does not implement all ops on MPS (suppressed when `PYTORCH_ENABLE_MPS_FALLBACK=1`).

## Decision

Match `sbi`: **default `device = "cpu"`**, GPU is opt-in.

This is consistent with the profiling above, where MPS is slower than CPU for the small nets in the SIR-scale examples. Defaulting everyone to GPU would regress the common case. Plan:

- `npe()` / `nle()` gain `device = "cpu"`.
- Accept `"cpu"`, `"cuda"`, `"mps"`, and a `"gpu"`/`"auto"` string that resolves CUDA -> MPS -> CPU (mirroring `sbi`'s `"gpu"`).
- Validate portably: `"cuda"` only if `cuda_is_available()`, `"mps"` only if `backends_mps_is_available()`; error (or warn + CPU fallback) otherwise. One-time MPS caveat note.
- Then do the device-safe tensor plumbing described in the issue body so a non-CPU device actually works end to end.
